### ISSUE TITLE
memory.c:286:2: error: #error No suitable alligned alloc found, please report to john-dev mailing list (state your OS details).

### ISSUE THREAD

**Issue by huanghemengxi** [magnumripper/JohnTheRipper#2153]
Hello everyone, I am in compiling JohnTheRipper-bleeding-jumbo version of the program.
Error prompt: memory.c: In function 'mem_alloc_align_func':
memory.c: 286: 2: error: #error No suitable alligned alloc found, please report to john-dev mailing list (state your OS det
ails).
make [1]: *** [memory.o] Error 1
make [1]: Leaving directory `/ opt / Johnb / src '
make: *** [default] Error 2
Environment configuration is as follows:
Operating systems: AIX Version 5.2
make version: GNU Make 3.81
gcc version: gcc version 4.6.1 (GCC)

**Comment by magnumripper** [comment]
So AIX 5.2 have neither of posix_memalign(), aligned_alloc() or memalign(). On the other hand it'd probably do just fine with a plain alloc() here since we won't use any SIMD or stuff like that... perhaps we should just fall back to that with a warning?

**Comment by solardiz** [comment]
> we won't use any SIMD or stuff like that...

Why not? AIX runs on POWER hardware, and we have SIMD (AltiVec) code for that. Ideally, we should use that code. Maybe malloc() on AIX provides sufficient (16-byte) alignment for AltiVec, or maybe not - I don't know.

**Comment by magnumripper** [comment]
It seems to state the usual *"The block is aligned so that it can be used for any type of data"* but as long as Altivec is no wider than 16 that should be fine.

**Comment by solardiz** [comment]
> "The block is aligned so that it can be used for any type of data"

Yet on other systems this statement does not always apply to anything wider than "double", so it might not be true for AltiVec on all/older AIX as well.

**Comment by magnumripper** [comment]
Well we could fall back to valloc() instead... that'd give us page alignment %-)

**Comment by solardiz** [comment]
> Well we could fall back to valloc() instead... that'd give us page alignment %-)

The scrypt code actually wants page alignment. As to the rest of jumbo, perhaps it should use our mem_alloc_tiny() for anything SIMD. mem_alloc_tiny() is capable of arbitrary alignment at the expense of not being able to free. Allocating/freeing stuff while cracking has performance cost and memory fragmentation risk anyway.

**Comment by magnumripper** [comment]
> As to the rest of jumbo, perhaps it should use our mem_alloc_tiny() for anything SIMD

IIRC we had problems when running things like all-formats-tests (several hundreds of formats) on systems running OpenMP with plenty of cores but not so much memory. But the OpenCL case is much much worse than OpenMP - I can't remember the details now.

> Allocating/freeing stuff while cracking has performance cost and memory fragmentation risk anyway.

We almost never do that within the inner loops and when we do, we usually don't need it aligned (off the top of my head).

Anyway, here's a half-crazy KISS solution that seems to work just fine (can't test it on AIX though):
```diff
diff --git a/src/memory.c b/src/memory.c
index 5a42f17..0027e89 100644
--- a/src/memory.c
+++ b/src/memory.c
@@ -283,7 +283,17 @@ void *mem_alloc_align_func(size_t size, size_t align
                pexit("_aligned_malloc ("Zu" bytes)", size);
 
 #elif AC_BUILT
-#error No suitable alligned alloc found, please report to john-dev mailing list (state your OS details).
+/*
+ * No aligned alloc found, using valloc when needed (this gives us
+ * alignment to page size)
+ */
+       if (!(ptr = malloc(size)))
+               pexit("aligned_alloc ("Zu" bytes)", size);
+       if ((size_t)ptr & (align - 1)) {
+               MEM_FREE(ptr);
+               if (!(ptr = valloc(size)))
+                       pexit("aligned_alloc ("Zu" bytes)", size);
+       }
 
 /* we need an aligned alloc function for legacy builds */
 #elif _ISOC11_SOURCE
```

**Comment by solardiz** [comment]
> IIRC we had problems when running things like all-formats-tests (several hundreds of formats) on systems running OpenMP with plenty of cores but not so much memory. But the OpenCL case is much much worse than OpenMP

Oh, you're right. Multi-format tests are the reason to be able to free.

> Anyway, here's a half-crazy KISS solution that seems to work just fine (can't test it on AIX though):

I'm afraid valloc() isn't always available, and where it is it might require feature macros to be defined to make it available. From the Linux man page: "The function valloc() appeared in 3.0BSD. It is documented as being obsolete in 4.3BSD, and as legacy in SUSv2. It does not appear in POSIX.1-2001."

**Comment by magnumripper** [comment]
> I'm afraid valloc() isn't always available, and where it is it might require feature macros to be defined to make it available.

Right, if actually committing something like that code we should add autoconf tests for valloc.

**Comment by huanghemengxi** [comment]
Hello everyone, I was to modify memory.c then during the test it?
+ / *
+ * No aligned alloc found, using valloc when needed (this gives us
+ * Alignment to page size)
+ * /
+ If (! (Ptr = malloc (size)))
+ Pexit ( "aligned_alloc (" Zu "bytes)", size);
+ If ((size_t) ptr & (align - 1)) {
+ MEM_FREE (ptr);
+ If (! (Ptr = valloc (size)))
+ Pexit ( "aligned_alloc (" Zu "bytes)", size);
+}

**Comment by magnumripper** [comment]
Sure, you could give it a try.

**Comment by huanghemengxi** [comment]
Hello, modified memory.c file compiled without prompting, but the emergence of new error
1, wherein the at least one parameter an error message in the 285 line pcap_freecode SIPdump.c file (& fp)
各位好，修改memory.c文件后在编译没有提示了，但出现新的错误提示
1、其中一个错误提示在SIPdump.c文件的285行pcap_freecode(&fp)少一个参数
gcc -DAC_BUILT  -DJOHN_powerpc -c -g -O2 -I/usr/local/include -DARCH_LITTLE_ENDIAN=0   -Wall -Wdeclaration-after-statemen
t -fomit-frame-pointer -Wno-deprecated-declarations  -Wunused-but-set-variable -std=gnu89 -D_FILE_OFFSET_BITS=64 -D_LARGE
FILE64_SOURCE   -I/opt/freeware/include     -D_THREAD_SAFE    -I/usr/local/include -funroll-loops SIPdump.c -o SIPdump.o
In file included from SIPdump.c:51:0:
tcphdr.h:20:10: error: expected ':', ',', ';', '}' or '__attribute__' before '.' token
SIPdump.c: In function 'main':
SIPdump.c:285:2: warning: passing argument 1 of 'pcap_freecode' from incompatible pointer type [enabled by default]
/usr/include/pcap.h:141:5: note: expected 'struct pcap_t *' but argument is of type 'struct bpf_program *'
SIPdump.c:285:2: error: too few arguments to function 'pcap_freecode'
/usr/include/pcap.h:141:5: note: declared here
SIPdump.c: In function 'sniff_logins':
SIPdump.c:556:20: error: invalid application of 'sizeof' to incomplete type 'struct ip6_hdr' 
SIPdump.c:557:20: error: dereferencing pointer to incomplete type
SIPdump.c:558:16: error: dereferencing pointer to incomplete type
SIPdump.c:559:30: error: dereferencing pointer to incomplete type
SIPdump.c:560:30: error: dereferencing pointer to incomplete type
SIPdump.c:587:23: error: 'const struct tcp_hdr' has no member named 'ip_ff'
SIPdump.c: In function 'pcap_dloff':
SIPdump.c:880:7: error: 'DLT_IEEE802_11_RADIO' undeclared (first use in this function)
SIPdump.c:880:7: note: each undeclared identifier is reported only once for each function it appears in
SIPdump.c:883:7: error: 'DLT_IEEE802_11' undeclared (first use in this function)
make[1]: *** [SIPdump.o] Error 1
make[1]: Leaving directory `/opt/Johnb/src'
make: *** [default] Error 2

**Comment by magnumripper** [comment]
I have no idea what is happening there. Unless you can solve this yourself (and submit a patch) you could likely just configure without pcap support.

**Comment by huanghemengxi** [comment]
After installing the latest package libpcap solve the problem, thank you, magnum ripper

**Comment by magnumripper** [comment]
Thank you for reporting.

Note to self: According to http://www.gnu.org/software/autoconf-archive/ax_check_page_aligned_malloc.html:
*As an aside, note that valloc’d memory cannot safely be freed on all operating systems. (Again, some flavors of BSD are the troublemakers.) It’s best to avoid using valloc in favor of posix_memalign, memalign, or an aligned malloc as detected by AX_CHECK_PAGE_ALIGNED_MALLOC.*

**Comment by solardiz** [comment]
> tcphdr.h:20:10: error: expected ':', ',', ';', '}' or '**attribute**' before '.' token

This is still an issue reproducible on AIX. Yes, with old libpcap, but I think this one is unrelated to the rest. We have our own struct with different name from the system's - ours is `struct tcp_hdr`, the system's is `struct tcphdr`, so we kind of don't clash, but then we use the usual struct field names, and AIX has some of them as macros, so we clash.

AIX `/usr/include/netinet/tcp.h` relevant excerpt:

```c
#ifndef _TCPHDR
#define _TCPHDR
struct tcphdr {
        u_short th_sport;               /* source port */
        u_short th_dport;               /* destination port */
        tcp_seq th_seq;                 /* sequence number */
        tcp_seq th_ack;                 /* acknowledgement number */
        struct  ip_firstfour ip_ff;     /* see <netinet/ip.h> */
#define th_off          ip_v            /* offset */
#define th_x2           ip_hl           /* unused */
#define th_xoff         ip_vhl          /* offset+unused */
#define th_flags        ip_tos          /* flags */
#define th_win          ip_ff.ip_fwin   /* window (unsigned) */
#define TH_FIN  0x01
#define TH_SYN  0x02
#define TH_RST  0x04
#define TH_PUSH 0x08
#define TH_ACK  0x10
#define TH_URG  0x20
#define TH_ECN_ECHO  0x40               /* receiver received a CE packet */
#define TH_CWR  0x80                    /* Congestion Window Reduced (ECN) */
        u_short th_sum;                 /* checksum */
        u_short th_urp;                 /* urgent pointer */
};

typedef struct tcphdr tcphdr_t;
#endif /* _TCPHDR */
```

One way to fix this is:

```diff
+++ b/src/tcphdr.h
@@ -8,6 +8,11 @@ extern "C" {
 #include "arch.h"
 #include <stdint.h>
 
+#undef th_off
+#undef th_x2
+#undef th_flags
+#undef th_win
+
 struct tcp_hdr {
        uint16_t th_sport;
        uint16_t th_dport;
```

Another is:

```diff
+++ b/src/tcphdr.h
@@ -5,6 +5,10 @@
 extern "C" {
 #endif
 
+#ifdef _TCPHDR
+#define tcp_hdr tcphdr
+#else
+
 #include "arch.h"
 #include <stdint.h>
 
@@ -26,6 +30,8 @@ struct tcp_hdr {
        uint16_t th_urp;
 };
 
+#endif
+
 #ifdef __cplusplus
 }
 #endif
```

but I cannot easily fully test these since this system also has too old libpcap.

Also, we include `<getopt.h>` in 2 programs. This doesn't exist on AIX. The function is actually defined in `unistd.h`, but I suspect that on some other systems we could need to enable a feature flag to get it from there.

**Comment by solardiz** [comment]
I included some `struct tcphdr` and `<getopt.h>` fixes in a64f6cb3d01f7aa863309ff681df96dd6066c69b.
